Support for Opencv4 [WIP]

fmrico opened the pull request:
**Warning: This is the dirtiest PR I have ever made, but it's the only way that I have found to make it works in Ubuntu 20.04 with OpenCV4**

Hi All,

OpenCV 4 makes fail the compilation of darknet_ros for ROS Foxy. This PR makes it work. 

Issue referenced in #234 and #200, at least.

If someone finds a better way to compile this, **it is wellcome**. I hope this could help someone and inspire others for a better solution.

Working on ROS2 Foxy, Ubuntu 20.04 LTS and OpenCV 4.2.0

![darknert_ros_opencv4](https://user-images.githubusercontent.com/3810011/90510115-ae8cf100-e15a-11ea-8342-7144bb92d237.png)






Signed-off-by: Francisco Martin Rico <fmrico@gmail.com>

tomlankhorst commented:
Thanks, were you able to test on 18.04 too?

fmrico commented:
Hi @tomlankhorst 

No. The current version in the `ros2` branch (without this PR) works in 18.04/Eloquent. This PR tries to fix the significant change to 20.04 (GCC and OpenCV).

Maybe we could have an `eloquent` branch and a `foxy` branch, and redirect this PR to `foxy`. What do you think, @mbjelonic?

mbjelonic commented:
@fmrico Yes, we could have a seperate branch for your PR. @tomlankhorst What do you think?

tomlankhorst commented:
Let's start with a `foxy` branch, yes. Might eventually swap `foxy` for `master` and `master` for `melodic`.

fmrico commented:
Hi @tomasgareau 

Could you create the `foxy` branch from the `ros2` branch? It seems like it was created from `master`, and when I change the base branch, there is a lot of conflicts.

tomasgareau commented:
Hi @fmrico -- looks like you might have meant @tomlankhorst :)

fmrico commented:
Yes, sorry @tomasgareau . Fixing...

mbjelonic commented:
@fmrico I created a new branch called foxy and changed the base of this PR.

mbjelonic commented:
@fmrico Can you please update the readme and mention that this branch ist running with ROS2 Foxy, Ubuntu 20.04 LTS and OpenCV 4.2.0?

fmrico commented:
Hi @mbjelonic 

Thanks for creating the branch and changing the base!

I have updated the Readme. I also changed the compilation commands to use colcon.

ajaypaul2008 commented:
@fmrico I was trying to compile it with ros2 foxy and opencv 4.2.0 (output from the command `python3 -c "import cv2; print(cv2.__version__)"`) and still getting `/usr/include/opencv4/opencv2/core/cvdef.h:704:4: error: #error "OpenCV 4.x+ requires enabled C++11 support"
  704 | #  error "OpenCV 4.x+ requires enabled C++11 support"`
  Any idea what am I missing?

fmrico commented:
Hi @ajaypaul2008 

I thought that I fixed this with this PR :thinking: 

In any case, this PR was a horrible patch to make darknet_ros work with the decision of OpenCV. Maybe someone should do a real fix.

ajaypaul2008 commented:
@fmrico  Finally I got it working with opencv4 after I added changes from this PR https://github.com/leggedrobotics/darknet_ros/pull/202 to foxy branch in my machine
